File: Javascript/0278_First_Bad_Version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let l = 1;
        let r = n;
        let last;
        while(l <= r) {
            mid = Math.round((l+r)/2);
            if(isBadVersion(mid)) {
                r = mid-1;
            }
            else {
                l = mid+1;
            }
            
        }
        if(!isBadVersion(mid)) {
            return mid + 1;
        }
        return mid;
    };
};

File: Python/0278_First_Bad_Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 0, n
        while l <= r:
            mid = (l+r) // 2
            if isBadVersion(mid):
                r = mid - 1
            else:
                l = mid + 1
        if not isBadVersion(mid):
            return mid + 1
        return mid    
        